### PR TITLE
[LPT][CommonOptimizations] ShuffleChannelsTransformation and ShuffleChannelsFusion

### PR DESCRIPTION
**Pull request by v-Golubev**
Issues #44032, #44031

* LP Transformations: ShuffleChannelsTransformation implemented
* CommonOptimizations: ShuffleChannelsFusion implemented

Validation:
• Accuracy: developer-services/job/accuracy/1593/ (reference: developer-services/job/accuracy/1592/)
• Performance: developer-services/job/performance/796/ (reference: developer-services/job/performance/795/)

### CODE REVIEW

File: ngraph/core/include/ngraph/op/shuffle_channels.hpp [hunk 1]
```diff
@@ -34,7 +34,10 @@ namespace ngraph
             {
             public:
                 static constexpr NodeTypeInfo type_info{"ShuffleChannels", 0};
+                static constexpr NodeTypeInfo type_info_static{"ShuffleChannels", 0};
                 const NodeTypeInfo& get_type_info() const override { return type_info; }
+                static const NodeTypeInfo& get_type_info_static() { return type_info_static; }
```
Comment (eshoguli) [inline ngraph/core/include/ngraph/op/shuffle_channels.hpp:39]: use macros:
1. `NGRAPH_RTTI_DEFINITION(op::v1::ShuffleChannels, "ShuffleChannels", 0);` in implementation 
2. `NGRAPH_RTTI_DECLARATION` in declaration

Comment (v-Golubev) [reply ngraph/core/include/ngraph/op/shuffle_channels.hpp]: Thanks, done!

File: inference-engine/tests/functional/inference_engine/lp_transformations/shuffle_channles_transformation.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
+// Copyright (C) 2021 Intel Corporation
```
Comment (vzinovie) [inline inference-engine/tests/functional/inference_engine/lp_transformations/shuffle_channles_transformation.cpp:1]: Typo in file name

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/shuffle_channels_transformation.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#include "low_precision_transformations/shuffle_channels_transformation.hpp"
+#include "common_test_utils/test_constants.hpp"
+
+using namespace LayerTestsDefinitions;
+
+namespace {
+const std::vector<ngraph::element::Type> netPrecisions = {
+    ngraph::element::f32,
+    // ngraph::element::f16
```
Comment (vzinovie) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/shuffle_channels_transformation.cpp:15]: Why test on FP16 precision is turned off?

File: inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/shuffle_channels_function.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#include "lpt_ngraph_functions/common/builders.hpp"
+
+#include "lpt_ngraph_functions/shuffle_channels_function.hpp"
+#include "ngraph_functions/subgraph_builders.hpp"
+
+namespace ngraph {
+namespace builder {
+namespace subgraph {
+
+    std::shared_ptr<Function> ShuffleChannelsFunction::getOriginal(
```
Comment (vzinovie) [inline inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/shuffle_channels_function.cpp:17]: Remove indent

**Review by vzinovie — CHANGES_REQUESTED** [review]
Some minor changes required.

File: inference-engine/src/low_precision_transformations/src/shuffle_channels.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+#include <memory>
+#include <ngraph/ngraph.hpp>
+#include <ngraph/opsets/opset1.hpp>
+
+#include "low_precision/network_helper.hpp"
+
+namespace ngraph {
+namespace pass {
+namespace low_precision {
+    ShuffleChannelsTransformation::ShuffleChannelsTransformation(const Params& params) : LayerTransformation(params) {}
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/shuffle_channels.cpp]: Done, thanks!

File: inference-engine/src/low_precision_transformations/src/shuffle_channels.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        const auto shuffledMulConst = shuffleDequantizationConstant(dequantization.multiply);
+        replace_node(dequantization.multiplyConstant, shuffledMulConst);
+        dequantization.multiplyConstant = shuffledMulConst;
+
+        moveDequantizationAfter(context, shuffleChannels, dequantization, false);
+        return true;
+    }
+
+    bool ShuffleChannelsTransformation::canBeTransformed(const TransformationContext& context, std::shared_ptr<Node> op) const {
+        if (!LayerTransformation::canBeTransformedSpecialDimension(context, op)) {
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/shuffle_channels.cpp]: Done, thanks!

File: inference-engine/tests/functional/inference_engine/lp_transformations/shuffle_channles_transformation.cpp (hunk 2 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/inference_engine/lp_transformations/shuffle_channles_transformation.cpp]: Fixed, thanks!

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/shuffle_channels_transformation.cpp (hunk 3 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/shuffle_channels_transformation.cpp]: Fixed, thanks!

File: inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/shuffle_channels_function.cpp (hunk 4 above)
Comment (v-Golubev) [reply inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/shuffle_channels_function.cpp]: Done, thanks!

**Review by vzinovie — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+ngraph::pass::ShuffleChannelsFusion::ShuffleChannelsFusion(const bool reshape_constants_check) {
+    MATCHER_SCOPE(ShuffleChannelsFusion);
+    auto input0 = std::make_shared<pattern::op::Label>(element::f32);
+    auto input1 = std::make_shared<pattern::op::Label>(element::i64);
+    auto input2 = std::make_shared<pattern::op::Label>(element::i64);
+    auto input3 = std::make_shared<pattern::op::Label>(element::i64);
+    auto reshape_before = std::make_shared<ngraph::opset3::Reshape>(input0, input1, false);
+    auto transpose = std::make_shared<ngraph::opset3::Transpose>(reshape_before, input2);
+    auto reshape_after = std::make_shared<ngraph::opset3::Reshape>(transpose, input3, false);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:63]: 1. use `pattern::wrap_type` for pattern nodes
2. use `pattern::any_input` to specify node inputs if necessary
3. use operations from `opset6`

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    ngraph::matcher_pass_callback callback = [this, reshape_constants_check](pattern::Matcher& m) {
+        auto reshape_after = std::dynamic_pointer_cast<ngraph::opset3::Reshape>(m.get_match_root());
+        if (!reshape_after) {
+            return false;
+        }
+
+        auto transpose = std::dynamic_pointer_cast<ngraph::opset3::Transpose>(reshape_after->input_value(0).get_node_shared_ptr());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:71]: Do not traverse graph manually. Use `const auto& node_to_output = m.get_pattern_value_map();`

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        if (!transpose || transpose->get_output_target_inputs(0).size() != 1) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:72]: Use `pattern::consumers_count(1)` predicate to check number of consumers in pattern but not in callback.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            return false;
+        }
+
+        auto reshape_before = std::dynamic_pointer_cast<ngraph::opset3::Reshape>(transpose->input_value(0).get_node_shared_ptr());
+        if (!reshape_before || reshape_before->get_output_target_inputs(0).size() != 1) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:77]: the same.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            return false;
+        }
+
+        auto reshape_after_constant = std::dynamic_pointer_cast<ngraph::opset3::Constant>(reshape_after->input_value(1).get_node_shared_ptr());
+        auto transpose_constant = std::dynamic_pointer_cast<ngraph::opset3::Constant>(transpose->input_value(1).get_node_shared_ptr());
+        auto reshape_before_constant = std::dynamic_pointer_cast<ngraph::opset3::Constant>(reshape_before->input_value(1).get_node_shared_ptr());
+        if (!transpose_constant || !reshape_before_constant || !reshape_after_constant) {
+            return false;
+        }
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:86]: Move this check to pattern.
You need to pass `Constant` as input to Transpose and Reshape ops.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+        if (reshape_constants_check) {
+            auto reshape_before_values = reshape_before_constant->cast_vector<int64_t>();
+            auto reshape_after_values = reshape_after_constant->cast_vector<int64_t>();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:90]:
```suggestion
            const auto & reshape_before_values = reshape_before_constant->cast_vector<int64_t>();
            const auto & reshape_after_values = reshape_after_constant->cast_vector<int64_t>();
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+            for (size_t i = 0; i < reshape_before_values.size(); ++i) {
+                if ((reshape_before_values[i] == -1) || (reshape_before_values[i] == 0 && i != 0)) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:92]: Why we prohibit `0` value? Please explain.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        auto p_shape_reshape_before = reshape_before->get_output_partial_shape(0);
+        auto p_shape_transpose = transpose->get_output_partial_shape(0);
+        auto p_shape_transpose_constant = transpose_constant->get_output_partial_shape(0);
+        auto p_shape_reshape_after = reshape_after->get_output_partial_shape(0);
+
+        if (p_shape_input.is_dynamic() || p_shape_reshape_before.is_dynamic() ||
+            p_shape_transpose.is_dynamic() || p_shape_reshape_after.is_dynamic() ||
+            p_shape_transpose_constant.is_dynamic()) {
+            return false;
+        }
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:113]: move this check to pattern using `patter::has_static_shape()` predicate.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 7 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Done. Thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 8 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Done, thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 9 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Done, thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 10 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Done, thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 12 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Thanks, done

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 14 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Thanks, done

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 11 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Thanks, done!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 13 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: This is an unnecessary check. I removed it. Thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ShuffleChannelsFusion, "ShuffleChannelsFusion", 0);
+
+ngraph::pass::ShuffleChannelsFusion::ShuffleChannelsFusion(const bool reshape_constants_check) {
+    MATCHER_SCOPE(ShuffleChannelsFusion);
+    auto input = ngraph::pattern::any_input(pattern::has_static_shape());
+    auto reshape_before_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
+    auto transpose_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
+    auto reshape_after_const_pattern = ngraph::pattern::wrap_type<ngraph::opset6::Constant>(pattern::has_static_shape());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:61]: Constants always have static shapes, so you don't need to check it.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+    auto has_static_shape_and_single_consumer = [](const Output<Node>& output) {
+        return pattern::has_static_shape() && pattern::consumers_count(1);
+    };
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:65]: This predicates `pattern::has_static_shape` and `pattern::consumers_count(1)` return lambda function, so I assume that
 `&&` for two lambdas will always return true. You need to call this labmdas for an output like this:
```suggestion
    auto has_static_shape_and_single_consumer = [](const Output<Node>& output) {
        return pattern::has_static_shape()(output) && pattern::consumers_count(1)(output);
    };
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        if (!reshape_after || !transpose || !reshape_after)
+            return false;
+
+        if (reshape_constants_check) {
+            auto reshape_before_constant = std::dynamic_pointer_cast<ngraph::opset6::Constant>(
+                pattern_map.at(reshape_before_const_pattern).get_node_shared_ptr());
+            auto reshape_after_constant = std::dynamic_pointer_cast<ngraph::opset6::Constant>(
+                pattern_map.at(reshape_after_const_pattern).get_node_shared_ptr());
+
+            const auto& reshape_before_values = reshape_before_constant->cast_vector<int64_t>();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:89]: You always need to check that the result of dynamic cast is not nullptr.

File: inference-engine/tests/functional/inference_engine/transformations/shuffle_channels_fusion_test.cpp [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    os << " }";
+    return os;
+}
+
+class ShuffleChannelsFusion : public ::testing::Test, public testing::WithParamInterface<ShuffleChannelsFusionTestValues> {
+public:
+    void SetUp() override {
+        const auto values = GetParam();
+        {
+            auto input0 = std::make_shared<opset6::Parameter>(element::f32, Shape{ values.batch_size, 128, 720, 480 });
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/shuffle_channels_fusion_test.cpp:51]: Add tests with dynamic shape

**Comment by GlebKazantaev** [comment]
nGrpah transformation part looks good.

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 15 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Thanks, fixed!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 16 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp]: Thanks, fixed!

File: inference-engine/tests/functional/inference_engine/transformations/shuffle_channels_fusion_test.cpp (hunk 18 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/inference_engine/transformations/shuffle_channels_fusion_test.cpp]: Tests with dynamic shape added, thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp (hunk 17 above)
Comment (v-Golubev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/shuffle_channels_fusion.cpp:92]: Thanks, done

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**